File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/list/ClickItemListener.java
package com.lasthopesoftware.bluewater.client.library.items.list;

import android.content.Context;
import android.content.Intent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;

import com.lasthopesoftware.bluewater.client.connection.SessionConnection;
import com.lasthopesoftware.bluewater.client.library.items.Item;
import com.lasthopesoftware.bluewater.client.library.items.access.ItemProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.list.FileListActivity;

import java.util.ArrayList;

import static com.vedsoft.futures.callables.VoidFunc.runCarelessly;

public class ClickItemListener implements OnItemClickListener {

	private final ArrayList<Item> mItems;
	private final Context mContext;

	public ClickItemListener(Context context, ArrayList<Item> items) {
		mContext = context;
        mItems = items;
	}
	
	@Override
	public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        final Item item = mItems.get(position);

        ItemProvider.provide(SessionConnection.getSessionConnectionProvider(), item.getKey())
            .then(runCarelessly(items -> {
				if (items == null) return;

				if (items.size() > 0) {
					final Intent itemlistIntent = new Intent(mContext, ItemListActivity.class);
					itemlistIntent.putExtra(ItemListActivity.KEY, item.getKey());
					itemlistIntent.putExtra(ItemListActivity.VALUE, item.getValue());
					mContext.startActivity(itemlistIntent);

					return;
				}

				final Intent fileListIntent = new Intent(mContext, FileListActivity.class);
				fileListIntent.putExtra(FileListActivity.KEY, item.getKey());
				fileListIntent.putExtra(FileListActivity.VALUE, item.getValue());
				fileListIntent.setAction(FileListActivity.VIEW_ITEM_FILES);
				mContext.startActivity(fileListIntent);
			}));
	}

}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/list/ClickItemListener.kt
package com.lasthopesoftware.bluewater.client.browsing.items.list

import android.content.Intent
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import com.lasthopesoftware.bluewater.client.browsing.items.Item
import com.lasthopesoftware.bluewater.client.browsing.items.access.ItemProvider.Companion.promiseItems
import com.lasthopesoftware.bluewater.client.browsing.items.list.ItemListActivity
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.list.FileListActivity.Companion.startFileListActivity
import com.lasthopesoftware.bluewater.client.connection.session.SessionConnection.Companion.getInstance
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise
import org.joda.time.Duration
import org.slf4j.LoggerFactory

class ClickItemListener(private val items: List<Item>, private val loadingView: View) : OnItemClickListener {
	override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
		parent.visibility = ViewUtils.getVisibility(false)
		loadingView.visibility = ViewUtils.getVisibility(true)
		val item = items[position]
		val context = view.context
		getInstance(context).promiseSessionConnection()
			.eventually { c -> c.promiseItems(item.key) }
			.then({ items ->
				if (items.isNotEmpty()) {
					val itemListIntent = Intent(context, ItemListActivity::class.java)
					itemListIntent.putExtra(ItemListActivity.KEY, item.key)
					itemListIntent.putExtra(ItemListActivity.VALUE, item.value)
					context.startActivity(itemListIntent)
					return@then
				}
				startFileListActivity(context, item)
			}, { e -> logger.error("An error occurred getting nested items for item " + item.key, e) })
			.eventually {
				LoopedInPromise({
					parent.visibility = ViewUtils.getVisibility(true)
					loadingView.visibility = ViewUtils.getVisibility(false)
				}, context, Duration.standardSeconds(1))
			}
	}

	companion object {
		private val logger = LoggerFactory.getLogger(ClickItemListener::class.java)
	}
}
